File: RC_robot/RC_robot/arm_dynamics.m
function output = arm_dynamics(x, u, L)
    m1 = 0.5;
    m2 = 0.5;
    b1 = 0.5;
    b2 = 0.5;

    l1 = L(1);
    l2 = L(2);

    % input arguments
    torque = [u(1); u(2)];
    theta1 = x(1);
    theta2 = x(2);
    theta1dot = x(3);
    theta2dot = x(4);

    % inertia matrix
    M11 = (m1 + m2) * l1^2 + m2 * l2^2 + 2 * m2 * l1 * l2 * cos(theta2);
    M12 = m2 * l2^2 + m2 * l1 * l2 * cos(theta2);
    M21 = m2 * l2^2 + m2 * l1 * l2 * cos(theta2);
    M22 = m2 * l2^2;
    M = [M11 M12; M21 M22];

    % Coriolis force vector
    V1 = - m2 * l1 * l2 * (2 * theta1dot * theta2dot + theta2dot^2) * sin(theta2);
    V2 = m2 * l1 * l2 * theta1dot^2 * sin(theta2);
    V = [V1; V2];

    % frition (viscosity) vector
    F = [b1 * theta1dot; b2 * theta2dot];

    thetaddot = inv(M) * (torque - F - V);
    output = [theta1dot; theta2dot; thetaddot(1); thetaddot(2)];
end

File: reservoir_motor/arm_dynamics.cpp
#include "arm_dynamics.h"

Vector arm_dynamics(const Vector& x, const Vector& u, const DynamicsParameters& dynamicsParams) {
    // パラメータの取得
    double m1 = dynamicsParams.m1;
    double m2 = dynamicsParams.m2;
    double b1 = dynamicsParams.b1;
    double b2 = dynamicsParams.b2;
    double l1 = dynamicsParams.L[0];
    double l2 = dynamicsParams.L[1];

    // 入力引数
    Vector torque(2);
    torque << u(0), u(1);
    double theta1 = x(0);
    double theta2 = x(1);
    double theta1dot = x(2);
    double theta2dot = x(3);

    // 慣性行列 M
    double M11 = (m1 + m2) * l1 * l1 + m2 * l2 * l2 + 2 * m2 * l1 * l2 * std::cos(theta2);
    double M12 = m2 * l2 * l2 + m2 * l1 * l2 * std::cos(theta2);
    double M21 = M12;
    double M22 = m2 * l2 * l2;

    Eigen::Matrix2d M;
    M << M11, M12,
         M21, M22;

    // コリオリ力ベクトル V
    double V1 = -m2 * l1 * l2 * (2 * theta1dot * theta2dot + theta2dot * theta2dot) * std::sin(theta2);
    double V2 = m2 * l1 * l2 * theta1dot * theta1dot * std::sin(theta2);

    Vector V(2);
    V << V1, V2;

    // 摩擦ベクトル F
    Vector F(2);
    F << b1 * theta1dot, b2 * theta2dot;

    // 加速度 thetaddot の計算
    Vector thetaddot = M.inverse() * (torque - F - V);

    // 結果を結合
    Vector output(4);
    output << theta1dot, theta2dot, thetaddot(0), thetaddot(1);

    return output;
}
